src/backend/api/objects.py

```diff
@@ -97,3 +143,464 @@ async def get_object_list(
+    """
+    async with conn.conn() as s3:
+        try:
+            s3_res = await s3.get_object_tagging(
+                Bucket=bucket_name, Key=params.Key, VersionId=params.VersionId
+            )
+            res = parse_obj_as(List[Tag], s3_res["TagSet"])
+        except s3.exceptions.ClientError as err:
+            if err.response["Error"]["Code"] == "NoSuchTagSet":
+                res = []
+            else:
+                raise err
+    return res
+
+
+@router.put(
+    "/{bucket_name}/tags",
+    response_model=bool,
+    responses=s3gw_client_responses(),
+)
+async def set_object_tagging(
+    conn: S3GWClientDep, bucket_name: str, params: SetObjectTaggingRequest
+) -> bool:
+    """
+    See
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/put_object_tagging.html
+    """
+    async with conn.conn() as s3:
+        try:
+            params_raw = params.dict()
+            await s3.put_object_tagging(
+                Bucket=bucket_name,
+                Key=params.Key,
+                VersionId=params.VersionId,
+                Tagging={"TagSet": params_raw["TagSet"]},
+            )
+        except s3.exceptions.ClientError:
+            return False
+    return True
+
+
+@router.get(
+    "/{bucket_name}/retention",
+    response_model=Optional[ObjectLockRetentionTypeDef],
+    responses=s3gw_client_responses(),
+)
+async def get_object_retention(
+    conn: S3GWClientDep, bucket_name: str, params: ObjectRequest
+) -> Optional[ObjectLockRetentionTypeDef]:
+    """
+    See
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/get_object_retention.html
+    """
+    async with conn.conn() as s3:
+        try:
+            s3_res = await s3.get_object_retention(
+                Bucket=bucket_name, Key=params.Key, VersionId=params.VersionId
+            )
+            res = parse_obj_as(ObjectLockRetentionTypeDef, s3_res["Retention"])
+        except s3.exceptions.ClientError as err:
+            if (
+                err.response["Error"]["Code"]
+                == "ObjectLockConfigurationNotFoundError"
+            ):
+                res = None
+            else:
+                raise err
+    return res
+
+
+@router.get(
+    "/{bucket_name}/legal-hold",
+    response_model=Optional[ObjectLockLegalHold],
+    responses=s3gw_client_responses(),
+)
+async def get_object_legal_hold(
+    conn: S3GWClientDep, bucket_name: str, params: ObjectRequest
+) -> Optional[ObjectLockLegalHold]:
+    """
+    See
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/client/get_object_legal_hold.html
+    """
+    async with conn.conn() as s3:
+        try:
+            s3_res = await s3.get_object_legal_hold(
+                Bucket=bucket_name, Key=params.Key, VersionId=params.VersionId
+            )
+            res = ObjectLockLegalHold.parse_obj(s3_res["LegalHold"])
```

Comment:
Is there a chance `LegalHold` will not be in `s3_res`, other than in case of an error?

Reply:
I don't think so.